lib/pbench/test/unit/server/query_apis/test_datasets_publish.py

```diff
@@ -40,210 +16,180 @@ class TestDatasetsPublish(Commons):
-                        "_shards": {"total": 2, "successful": 2, "failed": 0},
-                        "_seq_no": 10,
-                        "_primary_term": 3,
-                        "status": 200,
-                    }
-                }
-                if first:
-                    first = False
-                    u = item["update"]
-                    u["error"] = {
-                        "type": "document_missing_exception",
-                        "reason": "[_doc][6]: document missing",
-                        "index_uuid": "aAsFqTI0Tc2W0LCWgPNrOA",
-                        "shard": "0",
-                        "index": index,
-                    }
-                    u["status"] = 400
-
-                items.append(item)
-
-        response_payload = {
-            "took": 16,
-            "errors": True,
-            "items": items,
-        }
-
-        response = query_api(
-            self.pbench_endpoint,
-            self.elastic_endpoint,
-            self.payload,
-            "",
-            HTTPStatus.INTERNAL_SERVER_ERROR,
-            json=response_payload,
-            headers={"Authorization": "Bearer " + pbench_token},
-        )
-        response_doc = response.get_json(force=True)
-        message = response_doc["message"]
-        assert message == (
-            "DatasetsPublish: the query postprocessor was unable to complete: "
-            "Postprocessing error returning 500: '1 of 4 Elasticsearch document UPDATE operations failed' "
-            "[{'document_missing_exception': 1, 'ok': 3}]"
+        self.fake_elastic(monkeypatch, get_document_map, True)
+
+        response = client.post(
+            f"{server_config.rest_uri}/datasets/publish",
+            headers={"authorization": f"Bearer {pbench_token}"},
+            json=self.PAYLOAD,
         )
-        summary = response_doc["data"]
-        assert summary == {"ok": 3, "document_missing_exception": 1}
 
-    @pytest.mark.parametrize(
-        "exceptions",
-        (
-            {
-                "exception": requests.exceptions.ConnectionError(),
-                "status": HTTPStatus.BAD_GATEWAY,
-            },
-            {
-                "exception": requests.exceptions.Timeout(),
-                "status": HTTPStatus.GATEWAY_TIMEOUT,
-            },
-            {
-                "exception": requests.exceptions.InvalidURL(),
-                "status": HTTPStatus.INTERNAL_SERVER_ERROR,
-            },
-            {"exception": Exception(), "status": HTTPStatus.INTERNAL_SERVER_ERROR},
-        ),
-    )
-    def test_http_exception(
-        self,
-        client,
-        server_config,
-        query_api,
-        exceptions,
-        user_ok,
-        find_template,
-        get_document_map,
-        pbench_token,
+        # Verify the report and status
+        assert response.status_code == HTTPStatus.INTERNAL_SERVER_ERROR
+        assert response.json["data"] == {"ok": 28, "KIDDING": 3}
```

Comment:
I think that you were right to try to maintain some abstraction from ES.  And, in this particular case, it's probable that the recipient of the failure won't be able to do anything about it; so the important thing will be to have good diagnostics in the log.  So, perhaps, yeah, going with just "OK" and "failed" for the response would be appropriate.  😕

Reply:
I'm OK with just summarizing "ok" and "failure"... the problem is that I think this makes logging the actual errors more critical since the client is totally insulated; and while I agree that logging each error response could be hard to read, I don't see another practical way to make failure diagnosable. I'm thinking the errors should be `info` or even `warning` instead of `debug` so we'll actually see them. It's not like we expect a lot in real operation, and it indicates something wrong with Elasticsearch that we need to diagnose.

Reply:
> It's not like we expect a lot in real operation, and it indicates something wrong with Elasticsearch that we need to diagnose.

It's true -- we don't _expect_ them -- but, when they show up, it's generally spectacular.  And, that's why I'm strongly recommending that we try to digest them intelligently, rather than logging them individually.

> I'm thinking the errors should be info or even warning instead of debug so we'll actually see them.

I concur.  (I would use whichever of `warn` or `error` is not associated with user actions -- I don't think there is (or should be) any way that a user could provoke these failures, so, when they show up, it should be a "big deal".)

> I think this makes logging the actual errors more critical

I concur.

> I don't see another practical way to make failure diagnosable.

In my previous life, we dealt with this by taking the message that we wanted to log (e.g., the detailed message provided by ES), and using it as a key into a `dict` storing a count as the value (I didn't know about `defaultdict(int)`, then).  The messages tended to be specific but not unique, so when we had catastrophe, it boiled down to relatively few messages (with really large incidence counts...).